File: pkg/tenet/validate.go
package tenet

import (
	"fmt"
	"regexp"
	"strings"
)

// validateDefinitions checks all definitions for type correctness and required fields.
// Accumulates all errors (non-blocking).
func (e *Engine) validateDefinitions() {
	for id, def := range e.schema.Definitions {
		if def == nil {
			continue
		}

		// Check required fields
		if def.Required {
			if def.Value == nil {
				e.addError(id, "", ErrMissingRequired, fmt.Sprintf("Required field '%s' is missing", id), "")
			} else if def.Type == "string" || def.Type == "select" {
				// Empty string is also considered "missing" for required string/select fields
				if strVal, ok := def.Value.(string); ok && strVal == "" {
					e.addError(id, "", ErrMissingRequired, fmt.Sprintf("Required field '%s' is missing", id), "")
				}
			}
		}

		// Validate type if value is present
		if def.Value != nil {
			e.validateType(id, def)
		}
	}
}

// validateType ensures a value matches its definition type and constraints.
// Array values are allowed — the declared type describes the element type,
// used by collection operators (some/all/none). Scalar validation is skipped for arrays.
func (e *Engine) validateType(id string, def *Definition) {
	value := def.Value

	// Skip scalar validation for array values (used with some/all/none operators)
	if isSlice(value) {
		return
	}

	switch def.Type {
	case "string":
		strVal, ok := value.(string)
		if !ok {
			e.addError(id, "", ErrTypeMismatch, fmt.Sprintf("Field '%s' must be a string", id), "")
			return
		}
		// Validate string length constraints
		e.validateStringConstraints(id, strVal, def)

	case "number", "currency":
		numVal, ok := toFloat(value)
		if !ok {
			e.addError(id, "", ErrTypeMismatch, fmt.Sprintf("Field '%s' must be a number", id), "")
			return
		}
		// Validate numeric range constraints
		e.validateNumericConstraints(id, numVal, def)

	case "boolean":
		if _, ok := value.(bool); !ok {
			e.addError(id, "", ErrTypeMismatch, fmt.Sprintf("Field '%s' must be a boolean", id), "")
		}

	case "select":
		// Validate that value is one of the allowed options
		strVal, ok := value.(string)
		if !ok {
			e.addError(id, "", ErrTypeMismatch, fmt.Sprintf("Field '%s' must be a string", id), "")
			return
		}
		if !e.isValidOption(strVal, def.Options) {
			e.addError(id, "", ErrConstraintViolation, fmt.Sprintf("Field '%s' value '%s' is not a valid option", id, strVal), "")
		}

	case "attestation":
		// Attestations must be boolean
		if _, ok := value.(bool); !ok {
			e.addError(id, "", ErrTypeMismatch, fmt.Sprintf("Attestation '%s' must be a boolean", id), "")
		}

	case "date":
		// Validate date format
		if _, ok := parseDate(value); !ok {
			e.addError(id, "", ErrTypeMismatch, fmt.Sprintf("Field '%s' must be a valid date", id), "")
		}
	}
}

// validateNumericConstraints checks min/max bounds for numeric values.
func (e *Engine) validateNumericConstraints(id string, value float64, def *Definition) {
	if def.Min != nil && value < *def.Min {
		e.addError(id, "", ErrConstraintViolation, fmt.Sprintf("Field '%s' value %.2f is below minimum %.2f", id, value, *def.Min), "")
	}
	if def.Max != nil && value > *def.Max {
		e.addError(id, "", ErrConstraintViolation, fmt.Sprintf("Field '%s' value %.2f exceeds maximum %.2f", id, value, *def.Max), "")
	}
}

// validateStringConstraints checks length and pattern constraints for strings.
func (e *Engine) validateStringConstraints(id string, value string, def *Definition) {
	if def.MinLength != nil && len(value) < *def.MinLength {
		e.addError(id, "", ErrConstraintViolation, fmt.Sprintf("Field '%s' is too short (minimum %d characters)", id, *def.MinLength), "")
	}
	if def.MaxLength != nil && len(value) > *def.MaxLength {
		e.addError(id, "", ErrConstraintViolation, fmt.Sprintf("Field '%s' is too long (maximum %d characters)", id, *def.MaxLength), "")
	}
	if def.Pattern != "" {
		re, err := regexp.Compile(def.Pattern)
		if err == nil && !re.MatchString(value) {
			e.addError(id, "", ErrConstraintViolation,
				fmt.Sprintf("Field '%s' does not match required pattern", id), "")
		}
	}
}

// checkAttestations ensures all required attestations are confirmed.
// Validates both legacy attestations in definitions and rich attestations.
func (e *Engine) checkAttestations() {
	// Check legacy attestations in definitions (simple type: attestation)
	for id, def := range e.schema.Definitions {
		if def == nil || def.Type != "attestation" {
			continue
		}
		if def.Required && def.Value != true {
			e.addError(id, "", ErrAttestationIncomplete, fmt.Sprintf("Required attestation '%s' not confirmed", id), "")
		}
	}

	// Check rich attestations
	for id, att := range e.schema.Attestations {
		if att == nil {
			continue
		}

		// Process on_sign if signed is true
		if att.Signed && att.OnSign != nil {
			e.applyAction(att.OnSign, "attestation_"+id, att.LawRef)
		}

		// Validate required attestations
		if att.Required {
			if !att.Signed {
				e.addError(id, "", ErrAttestationIncomplete, fmt.Sprintf("Required attestation '%s' not signed", id), att.LawRef)
			} else if att.Evidence == nil || att.Evidence.ProviderAuditID == "" {
				e.addError(id, "", ErrAttestationIncomplete, fmt.Sprintf("Attestation '%s' signed but missing evidence", id), att.LawRef)
			}
		}
	}
}

// isValidOption checks if a value is in the allowed options list.
func (e *Engine) isValidOption(value string, options []string) bool {
	if options == nil {
		return true // No restrictions
	}
	for _, opt := range options {
		if opt == value {
			return true
		}
	}
	return false
}

// determineStatus calculates the document status based on validation errors.
func (e *Engine) determineStatus() DocStatus {
	for _, err := range e.errors {
		if err.Kind == ErrTypeMismatch {
			return StatusInvalid
		}
	}
	for _, err := range e.errors {
		if err.Kind == ErrMissingRequired || err.Kind == ErrAttestationIncomplete {
			return StatusIncomplete
		}
	}
	for _, err := range e.errors {
		if err.Kind == ErrConstraintViolation {
			return StatusInvalid
		}
	}
	return StatusReady
}

// containsString is a simple substring check.
func containsString(s, substr string) bool {
	return strings.Contains(s, substr)
}

File: js/src/core/validate.ts
/**
 * Definition validation and status determination.
 * Validates types, constraints, and required fields.
 */

import type { EvalState, Definition, ErrorKind, DocStatus, Action } from './types.js';
import { toFloat, parseDate } from './operators.js';

/**
 * Add an error to the state's error list.
 */
export function addError(
    state: EvalState,
    fieldId: string,
    ruleId: string,
    kind: ErrorKind,
    message: string,
    lawRef?: string
): void {
    state.errors.push({
        field_id: fieldId || undefined,
        rule_id: ruleId || undefined,
        kind,
        message,
        law_ref: lawRef || undefined,
    });
}

/**
 * Check if a value is one of the allowed options.
 */
function isValidOption(value: string, options: string[] | undefined): boolean {
    if (!options) {
        return true; // No restrictions
    }
    return options.includes(value);
}

/**
 * Validate numeric constraints (min/max).
 */
function validateNumericConstraints(
    state: EvalState,
    id: string,
    value: number,
    def: Definition
): void {
    if (def.min !== undefined && value < def.min) {
        addError(state, id, '', 'constraint_violation', `Field '${id}' value ${value.toFixed(2)} is below minimum ${def.min.toFixed(2)}`);
    }
    if (def.max !== undefined && value > def.max) {
        addError(state, id, '', 'constraint_violation', `Field '${id}' value ${value.toFixed(2)} exceeds maximum ${def.max.toFixed(2)}`);
    }
}

/**
 * Validate string constraints (length, pattern).
 */
function validateStringConstraints(
    state: EvalState,
    id: string,
    value: string,
    def: Definition
): void {
    if (def.min_length !== undefined && value.length < def.min_length) {
        addError(state, id, '', 'constraint_violation', `Field '${id}' is too short (minimum ${def.min_length} characters)`);
    }
    if (def.max_length !== undefined && value.length > def.max_length) {
        addError(state, id, '', 'constraint_violation', `Field '${id}' is too long (maximum ${def.max_length} characters)`);
    }
    if (def.pattern) {
        try {
            const regex = new RegExp(def.pattern);
            if (!regex.test(value)) {
                addError(state, id, '', 'constraint_violation', `Field '${id}' does not match required pattern`);
            }
        } catch {
            // Invalid regex pattern, skip validation
        }
    }
}

/**
 * Validate a single definition's type and constraints.
 * Array values are allowed — the declared type describes the element type,
 * used by collection operators (some/all/none). Scalar validation is skipped for arrays.
 */
function validateType(state: EvalState, id: string, def: Definition): void {
    const value = def.value;

    // Skip scalar validation for array values (used with some/all/none operators)
    if (Array.isArray(value)) {
        return;
    }

    switch (def.type) {
        case 'string': {
            if (typeof value !== 'string') {
                addError(state, id, '', 'type_mismatch', `Field '${id}' must be a string`);
                return;
            }
            validateStringConstraints(state, id, value, def);
            break;
        }

        case 'number':
        case 'currency': {
            const [numVal, ok] = toFloat(value);
            if (!ok) {
                addError(state, id, '', 'type_mismatch', `Field '${id}' must be a number`);
                return;
            }
            validateNumericConstraints(state, id, numVal, def);
            break;
        }

        case 'boolean': {
            if (typeof value !== 'boolean') {
                addError(state, id, '', 'type_mismatch', `Field '${id}' must be a boolean`);
            }
            break;
        }

        case 'select': {
            if (typeof value !== 'string') {
                addError(state, id, '', 'type_mismatch', `Field '${id}' must be a string`);
                return;
            }
            if (!isValidOption(value, def.options)) {
                addError(state, id, '', 'constraint_violation', `Field '${id}' value '${value}' is not a valid option`);
            }
            break;
        }

        case 'attestation': {
            if (typeof value !== 'boolean') {
                addError(state, id, '', 'type_mismatch', `Attestation '${id}' must be a boolean`);
            }
            break;
        }

        case 'date': {
            const [, ok] = parseDate(value);
            if (!ok) {
                addError(state, id, '', 'type_mismatch', `Field '${id}' must be a valid date`);
            }
            break;
        }
    }
}

/**
 * Validate all definitions for type correctness and required fields.
 */
export function validateDefinitions(state: EvalState): void {
    for (const [id, def] of Object.entries(state.schema.definitions)) {
        if (!def) {
            continue;
        }

        // Check required fields
        if (def.required) {
            if (def.value === undefined || def.value === null) {
                addError(state, id, '', 'missing_required', `Required field '${id}' is missing`);
            } else if ((def.type === 'string' || def.type === 'select') && def.value === '') {
                // Empty string is also considered "missing" for required string/select fields
                addError(state, id, '', 'missing_required', `Required field '${id}' is missing`);
            }
        }

        // Validate type if value is present
        if (def.value !== undefined && def.value !== null) {
            validateType(state, id, def);
        }
    }
}

/**
 * Check attestations for required signatures.
 */
export function checkAttestations(
    state: EvalState,
    applyAction: (action: Action, ruleId: string, lawRef: string) => void
): void {
    // Check legacy attestations in definitions (simple type: attestation)
    for (const [id, def] of Object.entries(state.schema.definitions)) {
        if (!def || def.type !== 'attestation') {
            continue;
        }
        if (def.required && def.value !== true) {
            addError(state, id, '', 'attestation_incomplete', `Required attestation '${id}' not confirmed`);
        }
    }

    // Check rich attestations
    if (!state.schema.attestations) {
        return;
    }

    for (const [id, att] of Object.entries(state.schema.attestations)) {
        if (!att) {
            continue;
        }

        // Process on_sign if signed is true
        if (att.signed && att.on_sign) {
            applyAction(att.on_sign, `attestation_${id}`, att.law_ref || '');
        }

        // Validate required attestations
        if (att.required) {
            if (!att.signed) {
                addError(state, id, '', 'attestation_incomplete', `Required attestation '${id}' not signed`, att.law_ref);
            } else if (!att.evidence || !att.evidence.provider_audit_id) {
                addError(state, id, '', 'attestation_incomplete', `Attestation '${id}' signed but missing evidence`, att.law_ref);
            }
        }
    }
}

/**
 * Determine document status based on ErrorKind.
 * Non-blocking kinds (runtime_warning, cycle_detected, notice) do not affect status.
 */
export function determineStatus(state: EvalState): DocStatus {
    for (const err of state.errors) {
        if (err.kind === 'type_mismatch') return 'INVALID';
    }
    for (const err of state.errors) {
        if (err.kind === 'missing_required' || err.kind === 'attestation_incomplete') {
            return 'INCOMPLETE';
        }
    }
    for (const err of state.errors) {
        if (err.kind === 'constraint_violation') return 'INVALID';
    }
    return 'READY';
}
